Reorganize intro section on main page (#16226)
matplotlib/matplotlib: doc/index.rst

.. title:: Matplotlib: Python plotting

> Matplotlib: 2D visualization in Python

Reviewer: I know 2D is in the original copy, but technically mpl supports enough 3D and there's enough really cool stuff done in 3D that I don't know that 2d should be emphasized here

Author: ok

--------------------------------------

> Matplotlib supports your visualization tasks, from simple one-off plotting to

Reviewer 2: #
```suggestion
Matplotlib supports visualization tasks, from simple plots to
```

Author: Took "plot", kept "your"; I explicitly would like to address and include the user in this very first. It's not only what we do, but it addresses the user's needs.

> complex and interactive figures.

Reviewer 3: I think this is a very good summary of what Matplotlib does, but I would love to see the stress on "publication quality" figures in this one line summary.
It's all going to be about tradeoff, so this comment should (of course) not hold this PR. Here's a suggestion that builds on the work you've already done.

- Matplotlib supports your visualization tasks, from simple one-off plotting to complex, publication quality, and interactive figures.

Author: Thanks for the feedback. I've added your suggestion.

Reviewer 3: Thanks!

   </div>
   <span class="clear_screenshots"></span>

> 

Reviewer 4: I personally really like this sentence (yes, this specific wording) and would be sad to see it go -- it really embodies the matplotlib philsophy to me ("making hard things possible" is what distinguishes matplotlib from many "simpler" wrappers, and "making easy things easy" means that you don't have to fiddle with the GUI toolkit and that you can output to a bunch of formats natively, as opposed to e.g. web-only (js) libs).

Author: I understand your point. I've thought for a while if I should keep this sentence.

**TL;DR I don't think the sentence is helpful or even suitable. But if people are attached to it, I would be willing to leave it in for now.**

I have three issues with it:

- "Matplotlib tries to ..." is too defensive. In a philosophy you don't try. If anything this would need to be

    Matplotlib makes easy things easy and hard things possible.

- Easy and hard things are highly vague and subjective.
  > "making easy things easy" means that you don't have to fiddle with the GUI toolkit and that you can output to a bunch of formats natively, as opposed to e.g. web-only (js) libs

  is your particular interpretation and focused on the output aspect.
  I've read it more broadly as "Things that I can easily describe should be easy to code."

  In it's generality, you could apply the statement to a lot of software (It appears to coined on the design of Perl). It's not a distinctive feature of matplotlib. I assume that most of the visualization libraries would claim that paradigm (with different interpretations of easy and hard).

- From a user perspective, IMHO we are not regarded as particularly simple. This is partly because we see core matplotlib as fundamental building blocks, and we have a limited number of high-level helpers to arrange them. This is a design choice, which is perfectly fine. I just don't think we should over-emphasize easy.

  It's not easy if I need several lines of code to make a simple plot with axes labels, title and legend. It's not easy if I have to calculate positions for [grouped bar charts](https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/barchart.html). There are higher-level libraries that can claim "simple" much more.

Therefore, instead of "easy things easy and hard things possible", I hope to covey the two aspects more clearly and accurately with:
- Easy creation of numerous standard plots
- Customizable and extendable

Reviewer 4: I'm not going to try to argue rationally here -- it is strictly a sentimental attachment to this sentence.  On the other hand, now that I stated my point, I'm not going to block its removal if there's agreement to do so.

(FWIW "Matplotlib makes easy things easy and hard things possible." is fine with me too, or perhaps "aims to ...".)

Reviewer: I'm with @anntzer on this - "Matplotlib makes easy things easy and hard things possible" is the brand/mission/ideal. And yes many users may not agree that we make easy things easy but this is the sales pitch. I think your text with some copy editing underneath works as an explanatory thing.

Author: Ok, I would like to talk about the misson statement at some point. It is valuable to clarify that for ourselves and also to communicate it. - Will likely be a paragraph or two on a dedicated page.

For now, I will add the easy sentence back in.

   </div>
   <span class="clear_screenshots"></span>

> Matplotlib makes easy things easy and hard things possible.

Author: I disagree here. The first sentence should tell what the library is about. Looking at other libraries, this seems to be common practice:

- NumPy is the fundamental package for scientific computing with Python.
- pandas is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.
- Seaborn is a Python data visualization library based on matplotlib.
- Bokeh is an interactive visualization library for modern web browsers.
- ggplot2 is a system for declaratively creating graphics, based on The Grammar of Graphics.

> Matplotlib makes easy things easy and hard things possible.

Reviewer: In that vein, the phrasing should probably be something like "Matplotlib is a comprehensive library for for building static, animated, and interactive visualizations in Python" or from the grant
" Matplotlib is the fundamental visualization library for the scientific Python Ecosystem" but I'm not too hung up on this

> Matplotlib makes easy things easy and hard things possible.

Author: I've taken your first proposal, but with "library for creating ..." instead of "building". The latter sounds too technical and like a lot of effort. :smile:

.. container:: bullet-box-container

   .. container:: bullet-box

      Create

>       - Easy creation of numerous standard plots

Reviewer 2: ```suggestion
      - Creation of standard plots
```

>       - Publication quality output to various file formats

Reviewer 2: ```suggestion
      - Publication quality output to standard file formats
```

Author: Took "common" instead. The sentence before already had "standard" and it feels repetitive otherwise.

        (png, pdf, svg, ps, ...)

>       - Interactive figures (zoom, pan, ...)

Reviewer: standard plots is too domain specific. I think instead stealing a bit from the next line: (and trying to keep all 3 boxes in parallel structure)
```
- Develop [publication quality plots](https://matplotlib.org/gallery/index.html) with just with just a few lines of code. 
- Build [unteractive figures](https://matplotlib.org/gallery/index.html#event-handling) that can zoom, pan, update...

Author: ok

   .. container:: bullet-box

      Customize

>       - Adapt and tune all aspects of your plot

Reviewer 2: ```suggestion
      - Adapt and tune all aspects of a graphic
```

      - Embed in interactive environments (e.g. IPython_, Jupyter_)

>       - Create graphics that respond to user GUI events

Reviewer: My reasoning here is to have the sentences flow under the subheading, and to link out to docs so this acts like a navigation hint of sorts - like these links are why these docs are telling you these things. The GUI stuff is already mentioned under `create`. Also stealing some copy that's already on the front page. 
```
- [Take full control](https://matplotlib.org/tutorials/index.html#tutorials) of line styles, font properties, axes properties...
- [Export and embed](https://matplotlib.org/api/index_backend_api.html) to a number of file formats and interactive environments
```
sorry not sure how these get formatted in rst form

Author: ok

>       - Respond to user actions based on the event system

Reviewer 2: ```suggestion
      - Create graphics that respond to user GUI events
```

      .. _IPython: http://ipython.org
      .. _Jupyter: http://jupyter.org

   .. container:: bullet-box

      Extend

      - Use additional tailored functionality through
        :doc:`third party packages <thirdpartypackages/index>`
      - Embed in GUI applications (Qt, Tk, Gtk3, Wx)

>       - Serve from web servers

Reviewer: Interactive, GUI stuff is already mentioned in create. And I have no attachment to that last bullet, was just trying to pull from stuff on the homepage and not repeat bullets. 
```
- Explore tailored functionality provided by :doc:`third party packages <thirdpartypackages/index>`
- Learn more about matplotlib through our many [external resources](https://matplotlib.org/resources/index.html)
```

Author: ok

>       - Learn more about matplotlib through our many

Reviewer: `the` instead of `our`

Author: changed.
